File: src/main/java/com/wgzhao/addax/admin/service/CollectionSchedulingService.java
package com.wgzhao.addax.admin.service;

import com.wgzhao.addax.admin.event.SourceUpdatedEvent;
import com.wgzhao.addax.admin.model.EtlSource;
import com.wgzhao.addax.admin.repository.EtlSourceRepo;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;

import java.time.LocalTime;
import java.util.List;

@Service
@Slf4j
public class CollectionSchedulingService {

    @Autowired
    private TaskSchedulerService taskSchedulerService;

    @Autowired
    private EtlSourceRepo etlSourceRepo;

    @Autowired
    private TaskService taskService;

    @Autowired
    private SystemConfigService systemConfigService;

    @EventListener(ApplicationReadyEvent.class)
    public void onApplicationReady() {
        rescheduleAllTasks();
    }

    public void rescheduleAllTasks() {
        try {
            // Schedule collection tasks for each source
            List<EtlSource> sources = etlSourceRepo.findByEnabled(true);
            for (EtlSource source : sources) {
                scheduleOrUpdateTask(source);
            }
            // Schedule the daily parameter update task
            scheduleDailyParamUpdate();
        } catch (Exception e) {
            // 记录详细异常，便于排查
           log.error("Error in rescheduleAllTasks: " , e);
        }
    }

    private void scheduleDailyParamUpdate() {
        LocalTime time = systemConfigService.getSwitchTimeAsTime();
        String cronExpression = convertLocalTimeToCron(time);
        Runnable task = taskService::updateParams;
        taskSchedulerService.scheduleTask("dailyParamUpdate", task, cronExpression);
    }

    public void scheduleOrUpdateTask(EtlSource source) {
        String taskId = "source-" + source.getCode();
        if (source.isEnabled() && source.getStartAt() != null) {
            log.info("Scheduling task for source {} at {}", source.getCode(), source.getStartAt());
            String cronExpression = convertLocalTimeToCron(source.getStartAt());
            // cancel existing task if any
            taskSchedulerService.cancelTask(taskId);
            Runnable task = () -> taskService.executeTasksForSource(source.getId());
            taskSchedulerService.scheduleTask(taskId, task, cronExpression);
        } else {
            taskSchedulerService.cancelTask(taskId);
        }
    }

    public void cancelTask(String code) {
        String taskId = "source-" + code;
        taskSchedulerService.cancelTask(taskId);
    }

    private String convertLocalTimeToCron(LocalTime time) {
        return String.format("0 %d %d * * ?", time.getMinute(), time.getHour());
    }

    @EventListener
    public void handleSourceEvent(SourceUpdatedEvent event) {
        if (event.isScheduleChanged()) {
            etlSourceRepo.findById(event.getSourceId()).ifPresent(this::scheduleOrUpdateTask);
        }
    }
}

File: src/main/kotlin/com/wgzhao/addax/admin/service/CollectionSchedulingService.kt
package com.wgzhao.addax.admin.service

import com.wgzhao.addax.admin.event.SourceUpdatedEvent
import com.wgzhao.addax.admin.model.EtlSource
import com.wgzhao.addax.admin.repository.EtlSourceRepo
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.boot.context.event.ApplicationReadyEvent
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Service
import java.time.LocalTime
import java.util.function.Consumer

@Service
class CollectionSchedulingService(
    private val taskSchedulerService: TaskSchedulerService,
    private val etlSourceRepo: EtlSourceRepo,
    private val taskService: TaskService,
    private val systemConfigService: SystemConfigService
    ) {

    private val log = KotlinLogging.logger {}

    @EventListener(ApplicationReadyEvent::class)
    fun onApplicationReady() {
        rescheduleAllTasks()
    }

    fun rescheduleAllTasks() {
        try {
            // Schedule collection tasks for each source
            val sources: MutableList<EtlSource>? = etlSourceRepo.findByEnabled(true)
            for (source in sources!!) {
                scheduleOrUpdateTask(source)
            }
            // Schedule the daily parameter update task
            scheduleDailyParamUpdate()
        } catch (e: Exception) {
            // 记录详细异常，便于排查
            log.error(e) { "Error in rescheduleAllTasks" }
        }
    }

    private fun scheduleDailyParamUpdate() {
        val time = systemConfigService.switchTimeAsTime
        val cronExpression = convertLocalTimeToCron(time)
        val task = Runnable { taskService.updateParams() }
        taskSchedulerService.scheduleTask("dailyParamUpdate", task, cronExpression)
    }

    fun scheduleOrUpdateTask(source: EtlSource) {
        val taskId = "source-${source.code}"
        if (source.enabled && source.startAt != null) {
            log.info {"Scheduling task for source ${source.code} at ${source.startAt}" }
            val cronExpression = convertLocalTimeToCron(source.startAt!!)
            // cancel existing task if any
            taskSchedulerService.cancelTask(taskId)
            val task = Runnable { taskService.executeTasksForSource(source.id) }
            taskSchedulerService.scheduleTask(taskId, task, cronExpression)
        } else {
            taskSchedulerService.cancelTask(taskId)
        }
    }

    fun cancelTask(code: String) {
        taskSchedulerService.cancelTask("source-${code}")
    }

    private fun convertLocalTimeToCron(time: LocalTime): String {
        return String.format("0 %d %d * * ?", time.minute, time.hour)
    }

    @EventListener
    fun handleSourceEvent(event: SourceUpdatedEvent) {
        if (event.connectionChanged) {
            etlSourceRepo.findById(event.sourceId).ifPresent(Consumer { source: EtlSource? -> this.scheduleOrUpdateTask(source!!) })
        }
    }
}
